File: gaze/shared/src/main/scala/dev/ligature/gaze/Nibblers.scala
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package dev.ligature.gaze

import dev.ligature.gaze.Gaze
import scala.collection.mutable.ArrayBuffer
import scala.collection.mutable

/** A Nibbler that takes a single item.
  */
def take[I](toMatch: I): Nibbler[I, I] = { (gaze) =>
  {
    gaze.next() match {
      case Some(i) =>
        if (toMatch == i) { Some(List(toMatch)) }
        else { None }
      case None => None
    }
  }
}

/** A Nibbler that takes a single item if the condition passed is true.
  */
def takeCond[I](cond: (I) => Boolean): Nibbler[I, I] = { (gaze) =>
  {
    gaze.next() match {
      case Some(i) =>
        if (cond(i)) { Some(List(i)) }
        else { None }
      case None => None
    }
  }
}

def takeAll[I, O](
    nibblers: Nibbler[I, O]*
): Nibbler[I, O] = { (gaze: Gaze[I]) =>
  {
    val results = ArrayBuffer[O]()
    val res = nibblers.forall { nibbler =>
      val res = gaze.attempt(nibbler)
      res match {
        case Some(res) => {
          results.appendAll(res)
          true
        }
        case None => {
          false
        }
      }
    }
    if (res) {
      Some(results.toSeq)
    } else {
      None
    }
  }
}

def takeAllGrouped[I](
    nibblers: Nibbler[I, I]*
): Nibbler[I, Seq[I]] = { (gaze: Gaze[I]) =>
  {
    val results = ArrayBuffer[Seq[I]]()
    val res = nibblers.forall { nibbler =>
      val res = gaze.attempt(nibbler)
      res match {
        case Some(res) => {
          results.append(res)
          true
        }
        case None => {
          false
        }
      }
    }
    if (res) {
      Some(results.toList)
    } else {
      None
    }
  }
}

def takeString(toMatch: String): Nibbler[Char, Char] = {
  //    let graphemes = to_match.graphemes(true).collect::<Vec<&str>>();
  val chars = toMatch.toVector
  return (gaze) => {
    var offset = 0
    var matched = true
    while (matched && offset < chars.length) {
      val nextChar = gaze.next()
      nextChar match {
        case Some(c) => {
          if (chars(offset) == c) {
            offset += 1;
          } else {
            matched = false
          }
        }
        case None => {
          matched = false
        }
      }
    }
    if (matched) {
      Some(chars)
    } else {
      None
    }
  }
}

def takeUntil[I](toMatch: I): Nibbler[I, I] = {
  return (gaze) => {
    val result = ArrayBuffer[I]()
    var matched = false
    while (!matched) {
      val next = gaze.peek()
      next match {
        case Some(v) => {
          if (v == toMatch) {
            matched = true
          } else {
            gaze.next()
            result.append(v)
          }
        }
        case None => {
          matched = true
        }
      }
    }
    Some(result.toSeq)
  }
}

def takeWhile[I](
    predicate: (toMatch: I) => Boolean
): Nibbler[I, I] = {
  return (gaze: Gaze[I]) => {
    val res = ArrayBuffer[I]()
    var matched = true
    var continue = true
    while (continue) {
      val peek = gaze.peek();

      peek match {
        case Some(c) => {
          if (predicate(c)) {
            gaze.next();
            res += c;
          } else if (res.length == 0) {
            matched = false
            continue = false
          } else {
            continue = false
          }
        }
        case None => {
          if (res.length == 0) {
            matched = false
            continue = false
          } else {
            continue = false
          }
        }
      }
    }
    if (matched) {
      Some(res.toSeq)
    } else {
      None
    }
  }
}

def optional[I, O](nibbler: Nibbler[I, O]): Nibbler[I, O] = { (gaze: Gaze[I]) =>
  {
    gaze.attempt(nibbler) match {
      case res: Some[_] => res
      case None         => Some(Seq())
    }
  }
}

def takeCharacters(chars: Char*): Nibbler[Char, Char] = takeWhile {
  chars.contains(_)
}

def takeFirst[I, O](
    nibblers: Nibbler[I, O]*
): Nibbler[I, O] = { (gaze: Gaze[I]) =>
  {
    var finalRes: Option[Seq[O]] = None
    val nibbler = nibblers.find { nibbler =>
      finalRes = gaze.attempt(nibbler)
      finalRes.isDefined
    }
    finalRes
  }
}

def repeat[I, O](
    nibbler: Nibbler[I, O]
): Nibbler[I, O] = { (gaze: Gaze[I]) =>
  {
    val allMatches = ArrayBuffer[O]()
    var continue = true
    while (!gaze.isComplete() && continue) {
      gaze.attempt(nibbler) match {
        case None    => continue = false
        case Some(v) => allMatches.appendAll(v)
      }
    }
    if (allMatches.isEmpty) {
      None
    } else {
      Some(allMatches.toSeq)
    }
  }
}

def between[I](
    wrapper: Nibbler[I, I],
    content: Nibbler[I, I]
) = takeAllGrouped(wrapper, content, wrapper).map(_(1))

def between[I](
    open: Nibbler[I, I],
    content: Nibbler[I, I],
    close: Nibbler[I, I]
) = takeAllGrouped(open, content, close).map(_(1))

File: gaze/src/commonMain/kotlin/dev/ligature/gaze/Nibblers.kt
/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

package dev.ligature.gaze

/** A Nibbler that takes a single item. */
fun <I> take(toMatch: I): Nibbler<I, I> = { gaze ->
  when (val next = gaze.next()) {
    null -> null
    else -> {
      if (toMatch == next) {
        listOf(toMatch)
      } else {
        null
      }
    }
  }
}

/** A Nibbler that takes a single item if the condition passed is true. */
fun <I> takeCond(cond: (I) -> Boolean): Nibbler<I, I> = { gaze ->
  when (val next = gaze.next()) {
    null -> null
    else ->
        if (cond(next)) {
          listOf(next)
        } else {
          null
        }
  }
}

fun <I, O> takeCondMap(cond: (I) -> O?): Nibbler<I, O> = { gaze ->
  when (val next = gaze.next()) {
    null -> null
    else -> {
      when (val res = cond(next)) {
        null -> null
        else -> listOf(res)
      }
    }
  }
}

/** A Nibbler that matches multiple Nibblers in order. */
fun <I, O> takeAll(vararg nibblers: Nibbler<I, O>): Nibbler<I, O> = { gaze: Gaze<I> ->
  val results = mutableListOf<O>()
  val res =
      nibblers.all { nibbler ->
        when (val res = gaze.attempt(nibbler)) {
          null -> false
          else -> {
            results.addAll(res)
            true
          }
        }
      }
  if (res) {
    results.toList()
  } else {
    null
  }
}

/** A Nibbler that matches multiple Nibblers in order but keeps each matching set in a List. */
fun <I, O> takeAllGrouped(vararg nibblers: Nibbler<I, O>): Nibbler<I, List<O>> = { gaze: Gaze<I> ->
  val results = mutableListOf<List<O>>()
  val res =
      nibblers.all { nibbler ->
        when (val res = gaze.attempt(nibbler)) {
          null -> false
          else -> {
            results.add(res)
            true
          }
        }
      }
  if (res) {
    results.toList()
  } else {
    null
  }
}

fun takeString(toMatch: String): Nibbler<Char, Char> {
  //    let graphemes = to_match.graphemes(true).collect::<Vec<&str>>();
  val chars = toMatch.toList()
  return { gaze ->
    var offset = 0
    var matched = true
    while (matched && offset < chars.size) {
      when (val nextChar = gaze.next()) {
        null -> matched = false
        else ->
            if (chars[offset] == nextChar) {
              offset += 1
            } else {
              matched = false
            }
      }
    }
    if (matched) {
      chars
    } else {
      null
    }
  }
}

/** A Nibbler that takes until a specific element is matched. */
fun <I> takeUntil(toMatch: I): Nibbler<I, I> = { gaze ->
  val result = mutableListOf<I>()
  var matched = false
  while (!matched) {
    when (val next = gaze.peek()) {
      null -> matched = true
      else ->
          if (next == toMatch) {
            matched = true
          } else {
            gaze.next()
            result.add(next)
          }
    }
  }
  result.toList()
}

// TODO needs tests
fun <I> takeUntil(toMatch: Nibbler<I, I>): Nibbler<I, I> = { gaze ->
  val result = mutableListOf<I>()
  var matched = false
  while (!matched) {
    when (val next = gaze.peek()) {
      null -> matched = true
      else -> {
        when (gaze.check(toMatch)) {
          null -> {
            gaze.next()
            result.add(next)
          }
          else -> matched = true
        }
      }
    }
  }
  result.toList()
}

fun <I> takeWhile(predicate: (toMatch: I) -> Boolean): Nibbler<I, I> = { gaze: Gaze<I> ->
  val res = mutableListOf<I>()
  var matched = true
  var proceed = true
  while (proceed) {
    when (val peek = gaze.peek()) {
      null ->
          if (res.size == 0) {
            matched = false
            proceed = false
          } else {
            proceed = false
          }
      else -> {
        if (predicate(peek)) {
          gaze.next()
          res += peek
        } else if (res.size == 0) {
          matched = false
          proceed = false
        } else {
          proceed = false
        }
      }
    }
  }
  if (matched) {
    res.toList()
  } else {
    null
  }
}

fun <I, O> optional(nibbler: Nibbler<I, O>): Nibbler<I, O> = { gaze: Gaze<I> ->
  when (val res = gaze.attempt(nibbler)) {
    null -> listOf()
    else -> res
  }
}

fun takeCharacters(vararg chars: Char): Nibbler<Char, Char> = takeWhile { chars.contains(it) }

fun <I, O> takeFirst(vararg nibblers: Nibbler<I, O>): Nibbler<I, O> = { gaze: Gaze<I> ->
  var finalRes: List<O>? = null
  nibblers.find { nibbler ->
    finalRes = gaze.attempt(nibbler)
    finalRes != null
  }
  finalRes
}

fun <I, O> repeat(nibbler: Nibbler<I, O>): Nibbler<I, O> = { gaze: Gaze<I> ->
  val allMatches = mutableListOf<O>()
  var proceed = true
  while (!gaze.isComplete && proceed) when (val v = gaze.attempt(nibbler)) {
    null -> proceed = false
    else -> allMatches.addAll(v)
  }
  if (allMatches.isEmpty()) {
    null
  } else {
    allMatches.toList()
  }
}

fun <I, O> between(wrapper: Nibbler<I, O>, content: Nibbler<I, O>) =
    takeAllGrouped(wrapper, content, wrapper).map { it[1] }

fun <I, O> between(open: Nibbler<I, O>, content: Nibbler<I, O>, close: Nibbler<I, O>) =
    takeAllGrouped(open, content, close).map { it[1] }
